### PR TITLE
[FEATURE] Ratex Fully Sharded Data Parallel ZeRO-1 Interface

### PR DESCRIPTION
**Pull request by shawnakdeb**
## Description ##
Created the FSDP style ZeRO-1 model wrapper. The goal of this module is to implement ZeRO-1 with a convenient interface that does not require updates to individual optimizers

cc @awslabs/raf-reviewer, @zachzzc, @zhen-jia, @XinweiFu

### CODE REVIEW

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+from raf import distributed as dist
+
+import ratex.optimizer as optim
+from ratex.core.lazy_model import all_gather, all_reduce, REDUCE_SUM
+
+
+class RatexFullyShardedDataParallel(nn.Module):
+    """FSDP ZeRO-1 wrapper"""
+
+    def __init__(self, module, optimizer=optim.SGD, optimizer_config=None):
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:17]: 1. Add docstring with argument types.
2. Add arguments types to all member functions.
Ref: https://github.com/awslabs/ratex/blob/main/ratex/core/lazy_model.py

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        self._shard_parameters()
+        # Optimizer initialization
+        self.optimizer = optimizer(self.sharded_parameters(), **optimizer_config or {})
+
+    def _shard_parameters(self):
+        """
+        Create list of this ranks shards
+        Shards are stored in a tuple with the respective parameter ie (parameter, shard)
+        """
+        self.sharded_params = []
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:35]: 1. Avoid creating class members other than constructor. You could put this line in `__init__` instead.
2. Why we need to shard parameters in ZeRO-1?

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        for param in self.params:
+            shard_data = self._get_shard(param.data)
+            shard = nn.Parameter(shard_data, requires_grad=param.requires_grad)
+            self.sharded_params.append(shard)
+
+    def _get_shard(self, tensor):
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:41]: This function is not informative. I'd suggest `_shard_tensor` or something like that.

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        Generator for the sharded parameters
+        """
+        yield from self.sharded_params
+
+    def forward(self, *args, **kwargs):
+        """
+        Calculate the output of the model using the wrapped module
+        """
+        outputs = self.module(*args, **kwargs)
+        return outputs
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:63]:
```suggestion
        return self.module(*args, **kwargs)
```

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    def step(self, *args, **kwargs):
+        """
+        Step the optimizer and update parameter weights
+        """
+        # Reduce full gradients across ranks
+        # Assign gradient shards to the respective parameter shards
+        for param_index in range(len(self.params)):
+            param, shard = self.params[param_index], self.sharded_params[param_index]
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:72]:
```suggestion
        for param, shard in zip(self.params, self.sharded_params):
```

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            if param.grad is not None:
+                all_reduce(REDUCE_SUM, [param.grad], scale=1.0 / self.world_size)
+                shard.grad = self._get_shard(param.grad)
+
+        # Step the wrapped optimizer
+        loss = self.optimizer.step(*args, **kwargs)
+
+        # All gather the new weights across the ranks and assign them to the full parameters
+        for param_index in range(len(self.params)):
+            param, shard = self.params[param_index], self.sharded_params[param_index]
```
Comment (comaniac) [inline ratex/core/distributed/ratex_fully_sharded_data_parallel.py:82]:
```suggestion
        for param, shard in zip(self.params, self.sharded_params):
```

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        epoch_loss = running_loss / dataset_sizes["train"]
+    return epoch_loss
+
+
+@pytest.mark.parametrize(
+    "optimizer", [(SGD, {"lr": 0.001, "momentum": 0.1}, 1), (Adam, {"lr": 0.001}, 2)]
+)
+def test_ratex_fully_sharded_data_parallelism_zero1(optimizer, tolerance=1e-10, seed=0):
+    torch.manual_seed(seed)
+    model_mnm = SingleLayerLogistics()
```
Comment (comaniac) [inline tests/python/model/test_ratex_fully_sharded_data_parallel.py:89]: This is a PyTorch model so shouldn't be named `mnm`?

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+        epoch_loss = running_loss / dataset_sizes["train"]
+    return epoch_loss
+
+
+@pytest.mark.parametrize(
+    "optimizer", [(SGD, {"lr": 0.001, "momentum": 0.1}, 1), (Adam, {"lr": 0.001}, 2)]
+)
+def test_ratex_fully_sharded_data_parallelism_zero1(optimizer, tolerance=1e-10, seed=0):
+    torch.manual_seed(seed)
```
Comment (comaniac) [inline tests/python/model/test_ratex_fully_sharded_data_parallel.py:88]: You can use `@with_seed` with `train`, so that every call to `train` should have the same random seed and you can remove all `manual_seed` in this function.

An alternative is making dataloader in this function and pass to `train`, since the traverse order is identical for every epoch whatever for the same or different models.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    total_rank, rank, local_rank = get_dist_comm_info()
+    device = lm.lazy_device(rank)
+    optimizer_zero1_loss = train(device, model_mnm, optimizer[0], optimizer[1], image_datasets)
+
+    torch.manual_seed(seed)
+    dcfg.zero_opt_level = 0
+    model_mnm = SingleLayerLogistics()
+    no_zero1_loss = train(device, model_mnm, optimizer[0], optimizer[1], image_datasets)
+
+    assert abs(no_zero1_loss - optimizer_zero1_loss) < tolerance
```
Comment (comaniac) [inline tests/python/model/test_ratex_fully_sharded_data_parallel.py:122]: We have utility for this already: `ratex.testing.check`.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    torch.manual_seed(seed)
+    model_mnm = SingleLayerLogistics()
+    fsdp_zero1_loss = train(
+        device, model_mnm, optimizer[0], optimizer[1], image_datasets, fsdp=True
+    )
+
+    assert abs(fsdp_zero1_loss - optimizer_zero1_loss) < tolerance
```
Comment (comaniac) [inline tests/python/model/test_ratex_fully_sharded_data_parallel.py:130]: Ditto

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            inputs.requires_grad = True
+            labels_one_hot = torch.tensor(np.eye(10, dtype=np.float32)[labels])
+            labels_one_hot = labels_one_hot.to(device)  # One-hot
+            optimizer.zero_grad()
+            outputs = model(inputs)
+            loss = -torch.sum(outputs * labels_one_hot) / inputs.size(0)
+            loss.backward()
+            optimizer.step()
+            lm.mark_step()
+            running_loss += loss.item() * inputs.size(0)
```
Comment (comaniac) [inline tests/python/model/test_ratex_fully_sharded_data_parallel.py:78]: Although it's not a big deal in a unit test, it's better to follow the best practice of lazy tensor. Specifically, in this line you materialize the loss by `.item()`. It creates a barrier at this point and prevents the overlapping between execution and graph tracing. See https://github.com/awslabs/ratex/blob/main/docs/overlap_trace_and_computation.md for details.

File: docs/FSDP_example.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (comaniac) [inline docs/FSDP_example.py:2]: We do not recommend to put an example script in `docs` without a document. The example script can be an auxiliary of a document. For example, https://github.com/awslabs/ratex/blob/main/docs/QUICKSTART.md uses `lenet.py`.

**Review by comaniac — CHANGES_REQUESTED** [review]

**Comment by comaniac** [comment]
btw, you may need to mark your test to prevent it from being ran under a single GPU environment. This should be the reason of the CI failure. See https://github.com/awslabs/ratex/blob/main/tests/python/model/test_distributed_models.py#L30 for reference.

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Module wrapper implementing ZeRO-1 in an FSDP style interface"""
```
[the remark this answers is not in the document]
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py:4]: I understand, I was following the naming convention used by PyTorch/xla where they prefixed everything with Xla. What is your opinion, does it make more sense to remove the ratex prefix for our files/classes?

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 7 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py:90]: Hmmm, I got the name from a script that Zach gave me, I will ask him about it separately.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 8 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py:89]: @with_seed does not seem to work for me. Perhaps it is because it does not say the pytorch seed? Do you think I should try to make my own decorator? Passing in data loader to train did not seem to work either without manually resetting the seed.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 7 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py:90]: Thanks, he told me that it refers to the old package name, I will remove the mnm suffix.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 11 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py]: I changed it so that the loss is not materialized until the end of each epoch.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 10 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py]: Changed to use the check function

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 9 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py]: Changed to use the check function

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 4 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: Fixed

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 5 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: Changed to use zip

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 6 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: Changed to use zip

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 3 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: Changed the method name to _shard_tensor

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 2 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: I changed the initialization of sharded_params to the constructor. The parameters are sharded in this implementation to be passed to the optimizer. This way each rank only stares shards of the optimizer states.

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 1 above)
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: I added arguments as needed according to our convention.

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py:4]: I see. Then I don't have a strong preference if this is intentional in order to differentiate the PyTorch FSDP.

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 2 above)
Comment (comaniac) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py]: I got your point. You use shard parameters to create optimizer, so that optimizer states are also sharded. The shard parameters are parts of the optimizer states.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 8 above)
Comment (comaniac) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py:91]: Would that work if we put `torch.manual_seed` into `with_seed`? We can do that if so; otherwise we could just keep the current implementation.

**Review by comaniac — APPROVED** [review]
LGTM. Only minor comments to be addressed.
cc @zhen-jia @zachzzc

**Review by zachzzc — COMMENTED** [review]
LGTM. Please add the test under ci/batch/task_python_distributed.py, otherwise test won't run

File: ratex/core/distributed/ratex_fully_sharded_data_parallel.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (shawnakdeb) [reply ratex/core/distributed/ratex_fully_sharded_data_parallel.py:4]: Ok, then I will resolve this as is.

File: tests/python/model/test_ratex_fully_sharded_data_parallel.py (hunk 8 above)
Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py:91]: I believe that would work. I will give it a try, I just was not sure if it was appropriate to change the current implementation.

Comment (shawnakdeb) [reply tests/python/model/test_ratex_fully_sharded_data_parallel.py]: I could not get it to work with with_seed for some reason. I am not sure if I am doing something incorrectly. I did however change the code so that torch.manual_seed is set and model is initialized in train, so that the code fragment is not duplicated

**Review by zhen-jia — APPROVED** [review]
Please address the comments and the CI issue. Otherwise LGTM

**Comment by comaniac** [comment]
Thanks @shawnakdeb @zachzzc @zhen-jia
